lib/tapioca/rbi/rewriters/merge_trees.rb

```diff
@@ -0,0 +1,589 @@
+          attr_reader :conflicts
+
+          sig { params(output: Tree, left_name: String, right_name: String).void }
+          def initialize(output, left_name: "left", right_name: "right")
+            super()
+            @tree = output
+            @index = T.let(output.index, Index)
+            @scope_stack = T.let([@tree], T::Array[Tree])
+            @left_name = left_name
+            @right_name = right_name
+            @conflicts = T.let([], T::Array[Conflict])
+          end
+
+          sig { override.params(node: T.nilable(Node)).void }
+          def visit(node)
+            return unless node
+
+            case node
+            when Scope
+              prev = previous_definition(node)
+
+              if prev.is_a?(Scope)
+                if node.compatible_with?(prev)
+                  prev.merge_with(node)
+                else
+                  make_conflict_scope(prev, node)
+                end
+                @scope_stack << prev
+              else
+                copy = node.dup_empty
+                current_scope << copy
+                @scope_stack << copy
+              end
+              visit_all(node.nodes)
+              @scope_stack.pop
+            when Tree
+              current_scope.merge_with(node)
+              visit_all(node.nodes)
+            when Indexable
+              prev = previous_definition(node)
+              if prev
+                if node.compatible_with?(prev)
+                  prev.merge_with(node)
+                else
+                  make_conflict_tree(prev, node)
+                end
+              else
+                current_scope << node.dup
+              end
+            end
+          end
+
+          private
+
+          sig { returns(Tree) }
+          def current_scope
+            T.must(@scope_stack.last)
+          end
+
+          sig { params(node: Node).returns(T.nilable(Node)) }
+          def previous_definition(node)
+            case node
+            when Indexable
+              node.index_ids.each do |id|
+                others = @index[id]
+                return others.last unless others.empty?
+              end
+            end
+            nil
+          end
+
+          sig { params(left: Scope, right: Scope).void }
+          def make_conflict_scope(left, right)
+            @conflicts << Conflict.new(left: left, right: right, left_name: @left_name, right_name: @right_name)
+            scope_conflict = ScopeConflict.new(left: left, right: right, left_name: @left_name, right_name: @right_name)
+            left.replace(scope_conflict)
+          end
+
+          sig { params(left: Node, right: Node).void }
+          def make_conflict_tree(left, right)
+            @conflicts << Conflict.new(left: left, right: right, left_name: @left_name, right_name: @right_name)
+            tree = left.parent_conflict_tree
+            unless tree
+              tree = ConflictTree.new(left_name: @left_name, right_name: @right_name)
```

Comment:
Is there way we can make the conflict resolution strategy configurable? For example, I would very much like us to use the bundler `/rbi` folder contents from gems, if they exist and then for Tapioca to merge its generated RBI onto that. However, I think I would like the Tapioca generated nodes to overwrite the bundled RBI nodes without needing any manual override. If the conflict resolution strategy was pluggable, I could easily create a `RightConflictTree` class that would always use the nodes from the right tree, and pass it in here.

Btw, I am not necessarily asking this to be configurable as part of this PR, just want to understand if we have plans on doing that.

Reply:
Yes it's the plan.

Follow ups will add ways to keep/discard definitions depending on the source as well as a way to annotate definitions that should be kept 👍